File: client/e2e.go
package client

// End-to-end encryption.
//
// The server's X-Encrypt-Password option encrypts at rest, but the server sees
// the plaintext and the password on the way through. This does not: the key is
// generated on the client, never leaves it, and travels in the URL *fragment*
// — the one part of a URL that browsers, proxies and access logs never see,
// because it is stripped before the request is sent.
//
// Wire format, identical in Go and in the browser (see web/src/lib/e2e.ts):
//
//	"STE1"                       4 bytes  magic and version
//	nonce prefix                 7 bytes  random
//	metadata length              2 bytes  big-endian, ciphertext incl. tag
//	chunk 0                    variable   AES-256-GCM(metadata JSON)
//	chunk 1..n                            AES-256-GCM(plaintext chunk)
//
// Each chunk is at most 64 KiB of plaintext plus a 16-byte tag. The nonce is
// the prefix, a big-endian chunk counter, and a final-chunk marker:
//
//	nonce = prefix[0:7] || counter[4] || (0x01 on the last chunk else 0x00)
//
// Folding the marker into the nonce is what makes truncation detectable: a
// stream cut short cannot produce a chunk that opens under a final nonce, so
// dropping the tail is an authentication failure rather than a silent partial
// file. This is the STREAM construction, as used by age.
//
// AES-GCM rather than ChaCha20-Poly1305 because WebCrypto implements it
// natively in every browser, and the browser has to be able to decrypt.

import (
	"bufio"
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/base64"
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"strings"
)

const (
	// e2eMagic identifies the format and its version.
	e2eMagic = "STE1"

	// e2eKeySize is 32 bytes: AES-256.
	e2eKeySize = 32

	// e2eNoncePrefixSize leaves room for the counter and final marker in a
	// 12-byte GCM nonce.
	e2eNoncePrefixSize = 7

	// e2eChunkSize is the plaintext size of one chunk. 64 KiB keeps the
	// working set small enough to stream a file of any size.
	e2eChunkSize = 64 * 1024

	// e2eTagSize is the AES-GCM authentication tag.
	e2eTagSize = 16

	// e2eMetaLenSize prefixes the variable-length metadata chunk. Without it
	// a reader cannot tell where the metadata ends and the payload begins.
	e2eMetaLenSize = 2

	e2eHeaderSize      = len(e2eMagic) + e2eNoncePrefixSize + e2eMetaLenSize
	e2eChunkCipherSize = e2eChunkSize + e2eTagSize
)

// ErrWrongKey means the payload did not authenticate: the wrong key, or a
// modified or truncated file. The two are deliberately indistinguishable.
var ErrWrongKey = errors.New("could not decrypt — wrong key, or the file was modified or truncated")

// E2EMetadata is the small header encrypted alongside the payload, so the real
// filename and type need not be exposed in the URL.
type E2EMetadata struct {
	Name string `json:"name"`
	Type string `json:"type,omitempty"`
}

// E2EKey is a content encryption key.
type E2EKey []byte

// NewE2EKey returns a fresh random key.
func NewE2EKey() (E2EKey, error) {
	key := make([]byte, e2eKeySize)
	if _, err := rand.Read(key); err != nil {
		return nil, fmt.Errorf("could not generate a key: %w", err)
	}
	return key, nil
}

// String renders the key for a URL fragment: unpadded base64url, so it stays
// URL-safe and copy-pasteable.
func (k E2EKey) String() string {
	return base64.RawURLEncoding.EncodeToString(k)
}

// ParseE2EKey reads a key from its fragment encoding.
func ParseE2EKey(s string) (E2EKey, error) {
	key, err := base64.RawURLEncoding.DecodeString(strings.TrimSpace(s))
	if err != nil {
		return nil, fmt.Errorf("malformed key: %w", err)
	}
	if len(key) != e2eKeySize {
		return nil, fmt.Errorf("malformed key: got %d bytes, want %d", len(key), e2eKeySize)
	}
	return key, nil
}

// FragmentKey extracts the key from a share URL's fragment, returning nil when
// there is none. Accepts both `#k=<key>` and a bare `#<key>`.
func FragmentKey(rawURL string) (E2EKey, error) {
	_, fragment, found := strings.Cut(rawURL, "#")
	if !found || fragment == "" {
		return nil, nil
	}

	fragment = strings.TrimPrefix(fragment, "k=")

	return ParseE2EKey(fragment)
}

// StripFragment returns the URL without its fragment. A fragment is never sent
// to a server, but strip it explicitly so the key cannot leak through a
// mis-built request.
func StripFragment(rawURL string) string {
	base, _, _ := strings.Cut(rawURL, "#")
	return base
}

func e2eNonce(prefix []byte, counter uint32, final bool) []byte {
	nonce := make([]byte, 12)
	copy(nonce, prefix)
	binary.BigEndian.PutUint32(nonce[e2eNoncePrefixSize:], counter)
	if final {
		nonce[11] = 1
	}
	return nonce
}

func e2eAEAD(key E2EKey) (cipher.AEAD, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	return cipher.NewGCM(block)
}

// E2EStream is one encryption of one file. Because the ciphertext is a pure
// function of the key, the nonce prefix and the plaintext, a stream can
// regenerate its own bytes from any offset — which is what makes an interrupted
// client-side-encrypted upload resumable rather than a restart from zero.
type E2EStream struct {
	Key    E2EKey
	Prefix []byte
	Meta   E2EMetadata
}

// NewE2EStream starts an encryption with a fresh random nonce prefix.
func NewE2EStream(key E2EKey, meta E2EMetadata) (*E2EStream, error) {
	prefix := make([]byte, e2eNoncePrefixSize)
	if _, err := rand.Read(prefix); err != nil {
		return nil, fmt.Errorf("could not generate a nonce: %w", err)
	}

	return NewE2EStreamWithPrefix(key, prefix, meta)
}

// NewE2EStreamWithPrefix rebuilds a stream from a prefix recorded earlier, so a
// resumed upload continues the ciphertext the server already holds instead of
// starting a different one.
func NewE2EStreamWithPrefix(key E2EKey, prefix []byte, meta E2EMetadata) (*E2EStream, error) {
	if len(key) != e2eKeySize {
		return nil, fmt.Errorf("malformed key: got %d bytes, want %d", len(key), e2eKeySize)
	}
	if len(prefix) != e2eNoncePrefixSize {
		return nil, fmt.Errorf("malformed nonce prefix: got %d bytes, want %d", len(prefix), e2eNoncePrefixSize)
	}

	return &E2EStream{Key: key, Prefix: prefix, Meta: meta}, nil
}

// header is the fixed part of the format plus the encrypted metadata chunk.
func (s *E2EStream) header() ([]byte, error) {
	aead, err := e2eAEAD(s.Key)
	if err != nil {
		return nil, err
	}

	metaJSON, err := json.Marshal(s.Meta)
	if err != nil {
		return nil, err
	}
	if len(metaJSON) > e2eChunkSize {
		return nil, errors.New("metadata is too large")
	}

	// Chunk 0 is the metadata; the payload starts at 1.
	sealedMeta := aead.Seal(nil, e2eNonce(s.Prefix, 0, false), metaJSON, nil)
	if len(sealedMeta) > 0xFFFF {
		return nil, errors.New("metadata is too large")
	}

	header := append([]byte(e2eMagic), s.Prefix...)
	header = binary.BigEndian.AppendUint16(header, uint16(len(sealedMeta)))

	return append(header, sealedMeta...), nil
}

// Size returns the exact ciphertext length for a plaintext of n bytes.
func (s *E2EStream) Size(n int64) (int64, error) {
	return E2EEncryptedSize(n, s.Meta)
}

// Reader returns the whole ciphertext, encrypting as it is consumed so memory
// use is one chunk regardless of file size.
func (s *E2EStream) Reader(plaintext io.Reader) (io.Reader, error) {
	header, err := s.header()
	if err != nil {
		return nil, err
	}

	return s.chunks(plaintext, header, 1)
}

// ReaderFrom returns the ciphertext from byte offset onwards. The plaintext is
// rewound to the chunk the offset falls in and the leading bytes of that chunk
// are discarded, so any offset works — not only chunk boundaries.
func (s *E2EStream) ReaderFrom(plaintext io.ReadSeeker, offset int64) (io.ReadCloser, error) {
	if offset < 0 {
		return nil, errors.New("negative offset")
	}

	header, err := s.header()
	if err != nil {
		return nil, err
	}

	headerLen := int64(len(header))

	var (
		counter     = uint32(1)
		plainOffset int64
		skip        = offset
	)

	if offset >= headerLen {
		index := (offset - headerLen) / e2eChunkCipherSize
		counter = uint32(index) + 1
		plainOffset = index * e2eChunkSize
		skip = (offset - headerLen) - index*e2eChunkCipherSize
		header = nil
	}

	if _, err := plaintext.Seek(plainOffset, io.SeekStart); err != nil {
		return nil, err
	}

	reader, err := s.chunks(plaintext, header, counter)
	if err != nil {
		return nil, err
	}

	if skip > 0 {
		if _, err := io.CopyN(io.Discard, reader, skip); err != nil {
			_ = reader.Close()
			return nil, err
		}
	}

	return reader, nil
}

// chunks writes header (which may be empty) and then every chunk from counter
// onwards into a pipe. Closing the returned reader stops the goroutine.
func (s *E2EStream) chunks(plaintext io.Reader, header []byte, counter uint32) (io.ReadCloser, error) {
	aead, err := e2eAEAD(s.Key)
	if err != nil {
		return nil, err
	}

	prefix := s.Prefix
	pr, pw := io.Pipe()

	go func() {
		if len(header) > 0 {
			if _, err := pw.Write(header); err != nil {
				_ = pw.CloseWithError(err)
				return
			}
		}

		buf := make([]byte, e2eChunkSize)
		reader := bufio.NewReaderSize(plaintext, e2eChunkSize)

		for ; ; counter++ {
			n, readErr := io.ReadFull(reader, buf)

			final := false
			switch {
			case errors.Is(readErr, io.EOF), errors.Is(readErr, io.ErrUnexpectedEOF):
				// Short or empty read: nothing follows.
				final = true
			case readErr != nil:
				_ = pw.CloseWithError(readErr)
				return
			default:
				// A full chunk may still be the last one; look ahead.
				if _, peekErr := reader.Peek(1); errors.Is(peekErr, io.EOF) {
					final = true
				}
			}

			sealed := aead.Seal(nil, e2eNonce(prefix, counter, final), buf[:n], nil)
			if _, err := pw.Write(sealed); err != nil {
				_ = pw.CloseWithError(err)
				return
			}

			if final {
				break
			}
		}

		_ = pw.Close()
	}()

	return pr, nil
}

// E2EEncrypt returns a reader over the encrypted stream. Encryption happens as
// the reader is consumed, so memory use is one chunk regardless of file size.
func E2EEncrypt(plaintext io.Reader, key E2EKey, meta E2EMetadata) (io.Reader, error) {
	stream, err := NewE2EStream(key, meta)
	if err != nil {
		return nil, err
	}

	return stream.Reader(plaintext)
}

// E2EEncryptedSize returns the exact ciphertext length for a plaintext of size
// n with the given metadata, so an upload can still declare a Content-Length.
func E2EEncryptedSize(n int64, meta E2EMetadata) (int64, error) {
	metaJSON, err := json.Marshal(meta)
	if err != nil {
		return 0, err
	}

	size := int64(e2eHeaderSize) + int64(len(metaJSON)+e2eTagSize)

	// A zero-length payload still produces one (empty, final) chunk.
	full := n / e2eChunkSize
	rest := n % e2eChunkSize

	size += full * int64(e2eChunkCipherSize)
	if rest > 0 || full == 0 {
		size += rest + int64(e2eTagSize)
	}

	return size, nil
}

// E2EDecrypt authenticates and decrypts a stream, returning its metadata and a
// reader over the plaintext.
func E2EDecrypt(ciphertext io.Reader, key E2EKey) (E2EMetadata, io.Reader, error) {
	var meta E2EMetadata

	aead, err := e2eAEAD(key)
	if err != nil {
		return meta, nil, err
	}

	reader := bufio.NewReaderSize(ciphertext, e2eChunkCipherSize)

	header := make([]byte, e2eHeaderSize)
	if _, err := io.ReadFull(reader, header); err != nil {
		return meta, nil, errors.New("not an end-to-end encrypted file")
	}
	if string(header[:len(e2eMagic)]) != e2eMagic {
		return meta, nil, errors.New("not an end-to-end encrypted file (or a newer format)")
	}

	prefix := header[len(e2eMagic) : len(e2eMagic)+e2eNoncePrefixSize]
	metaLen := binary.BigEndian.Uint16(header[len(e2eMagic)+e2eNoncePrefixSize:])

	// Decrypt the metadata up front, so a wrong key is reported before the
	// caller starts writing a file rather than halfway through it.
	metaBuf := make([]byte, metaLen)
	if _, err := io.ReadFull(reader, metaBuf); err != nil {
		return meta, nil, ErrWrongKey
	}

	metaJSON, err := aead.Open(nil, e2eNonce(prefix, 0, false), metaBuf, nil)
	if err != nil {
		return meta, nil, ErrWrongKey
	}
	if err := json.Unmarshal(metaJSON, &meta); err != nil {
		return meta, nil, ErrWrongKey
	}

	return meta, &e2eReader{aead: aead, prefix: prefix, reader: reader, counter: 1}, nil
}

// e2eReader decrypts chunk by chunk as it is read.
type e2eReader struct {
	aead    cipher.AEAD
	prefix  []byte
	reader  *bufio.Reader
	counter uint32

	pending []byte
	done    bool
	err     error
}

func (r *e2eReader) Read(p []byte) (int, error) {
	for len(r.pending) == 0 {
		if r.err != nil {
			return 0, r.err
		}
		if r.done {
			return 0, io.EOF
		}
		if err := r.next(); err != nil {
			r.err = err
			return 0, err
		}
	}

	n := copy(p, r.pending)
	r.pending = r.pending[n:]
	return n, nil
}

func (r *e2eReader) next() error {
	buf := make([]byte, e2eChunkCipherSize)
	n, readErr := io.ReadFull(r.reader, buf)

	final := false
	switch {
	case errors.Is(readErr, io.EOF):
		// The stream ended without a chunk carrying the final marker.
		return ErrWrongKey
	case errors.Is(readErr, io.ErrUnexpectedEOF):
		final = true
	case readErr != nil:
		return readErr
	default:
		if _, peekErr := r.reader.Peek(1); errors.Is(peekErr, io.EOF) {
			final = true
		}
	}

	plain, err := r.aead.Open(nil, e2eNonce(r.prefix, r.counter, final), buf[:n], nil)
	if err != nil {
		return ErrWrongKey
	}

	r.counter++
	r.pending = plain
	r.done = final

	return nil
}

File: web/src/lib/e2e.ts
// End-to-end encryption in the browser.
//
// Byte-for-byte compatible with the Go implementation in client/e2e.go — see
// that file for the format and the reasoning. In short:
//
//   "STE1"            4 bytes  magic and version
//   nonce prefix      7 bytes  random
//   metadata length   2 bytes  big-endian, ciphertext incl. tag
//   chunk 0         variable   AES-256-GCM(metadata JSON)
//   chunk 1..n                 AES-256-GCM(64 KiB of plaintext)
//
//   nonce = prefix || counter(uint32 BE) || (1 on the last chunk else 0)
//
// The key lives in the URL fragment, which browsers never put on the wire, so
// the server stores ciphertext it cannot read.

const MAGIC = "STE1";
const KEY_SIZE = 32;
const NONCE_PREFIX_SIZE = 7;
const META_LEN_SIZE = 2;
const CHUNK_SIZE = 64 * 1024;
const TAG_SIZE = 16;

const HEADER_SIZE = MAGIC.length + NONCE_PREFIX_SIZE + META_LEN_SIZE;
const CHUNK_CIPHER_SIZE = CHUNK_SIZE + TAG_SIZE;

export interface E2EMetadata {
  name: string;
  type?: string;
}

export class DecryptError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "DecryptError";
  }
}

/** A fresh 256-bit content key. */
export function generateKey(): Uint8Array {
  return crypto.getRandomValues(new Uint8Array(KEY_SIZE));
}

/** Unpadded base64url, matching Go's base64.RawURLEncoding. */
export function encodeKey(key: Uint8Array): string {
  let binary = "";
  for (const byte of key) binary += String.fromCharCode(byte);
  return btoa(binary).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

export function decodeKey(encoded: string): Uint8Array {
  const normalized = encoded.trim().replace(/-/g, "+").replace(/_/g, "/");
  const padded = normalized + "=".repeat((4 - (normalized.length % 4)) % 4);

  let binary: string;
  try {
    binary = atob(padded);
  } catch {
    throw new DecryptError("malformed key");
  }

  if (binary.length !== KEY_SIZE) {
    throw new DecryptError(`malformed key: got ${binary.length} bytes, want ${KEY_SIZE}`);
  }

  const key = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) key[i] = binary.charCodeAt(i);
  return key;
}

/**
 * Read the key out of a share URL's fragment. Accepts `#k=<key>` and a bare
 * `#<key>`; returns null when there is no fragment.
 */
export function keyFromFragment(fragment: string): Uint8Array | null {
  const raw = fragment.replace(/^#/, "").replace(/^k=/, "");
  if (!raw) return null;
  return decodeKey(raw);
}

function nonce(prefix: Uint8Array, counter: number, final: boolean): Uint8Array {
  const iv = new Uint8Array(12);
  iv.set(prefix, 0);
  new DataView(iv.buffer).setUint32(NONCE_PREFIX_SIZE, counter, false); // big-endian
  iv[11] = final ? 1 : 0;
  return iv;
}

async function importKey(key: Uint8Array): Promise<CryptoKey> {
  return crypto.subtle.importKey("raw", key as BufferSource, "AES-GCM", false, [
    "encrypt",
    "decrypt",
  ]);
}

async function seal(
  cryptoKey: CryptoKey,
  iv: Uint8Array,
  plaintext: Uint8Array
): Promise<Uint8Array> {
  const sealed = await crypto.subtle.encrypt(
    { name: "AES-GCM", iv: iv as BufferSource, tagLength: TAG_SIZE * 8 },
    cryptoKey,
    plaintext as BufferSource
  );
  return new Uint8Array(sealed);
}

async function open(
  cryptoKey: CryptoKey,
  iv: Uint8Array,
  ciphertext: Uint8Array
): Promise<Uint8Array> {
  try {
    const opened = await crypto.subtle.decrypt(
      { name: "AES-GCM", iv: iv as BufferSource, tagLength: TAG_SIZE * 8 },
      cryptoKey,
      ciphertext as BufferSource
    );
    return new Uint8Array(opened);
  } catch {
    throw new DecryptError("wrong key, or the file was modified or truncated");
  }
}

/**
 * Encrypt a blob. The result is assembled as a Blob rather than one buffer, so
 * the browser can spill it to disk instead of holding the whole file in memory.
 */
export async function encrypt(
  data: Blob,
  key: Uint8Array,
  meta: E2EMetadata,
  onProgress?: (done: number, total: number) => void
): Promise<Blob> {
  const cryptoKey = await importKey(key);
  const prefix = crypto.getRandomValues(new Uint8Array(NONCE_PREFIX_SIZE));

  const metaJSON = new TextEncoder().encode(JSON.stringify(meta));
  const sealedMeta = await seal(cryptoKey, nonce(prefix, 0, false), metaJSON);

  if (sealedMeta.length > 0xffff) throw new Error("metadata is too large");

  const header = new Uint8Array(HEADER_SIZE);
  header.set(new TextEncoder().encode(MAGIC), 0);
  header.set(prefix, MAGIC.length);
  new DataView(header.buffer).setUint16(MAGIC.length + NONCE_PREFIX_SIZE, sealedMeta.length, false);

  const parts: BlobPart[] = [header as BlobPart, sealedMeta as BlobPart];

  // A zero-length payload still produces one empty, final chunk, so that the
  // stream always ends with the final marker.
  const chunkCount = Math.max(1, Math.ceil(data.size / CHUNK_SIZE));

  for (let i = 0; i < chunkCount; i++) {
    const start = i * CHUNK_SIZE;
    const slice = new Uint8Array(await data.slice(start, start + CHUNK_SIZE).arrayBuffer());
    const final = i === chunkCount - 1;

    parts.push((await seal(cryptoKey, nonce(prefix, i + 1, final), slice)) as BlobPart);
    onProgress?.(Math.min(start + CHUNK_SIZE, data.size), data.size);
  }

  return new Blob(parts, { type: "application/octet-stream" });
}

/** The exact ciphertext length for a plaintext of `size` bytes. */
export function encryptedSize(size: number, meta: E2EMetadata): number {
  const metaLen = new TextEncoder().encode(JSON.stringify(meta)).length;
  const chunkCount = Math.max(1, Math.ceil(size / CHUNK_SIZE));
  return HEADER_SIZE + metaLen + TAG_SIZE + size + chunkCount * TAG_SIZE;
}

/**
 * Decrypt a whole ciphertext. Throws DecryptError on a wrong key, a modified
 * payload, or a truncated stream — all three are indistinguishable by design.
 */
export async function decrypt(
  ciphertext: Uint8Array,
  key: Uint8Array,
  onProgress?: (done: number, total: number) => void
): Promise<{ meta: E2EMetadata; blob: Blob }> {
  if (ciphertext.length < HEADER_SIZE) {
    throw new DecryptError("not an end-to-end encrypted file");
  }

  const magic = new TextDecoder().decode(ciphertext.subarray(0, MAGIC.length));
  if (magic !== MAGIC) {
    throw new DecryptError("not an end-to-end encrypted file (or a newer format)");
  }

  const cryptoKey = await importKey(key);
  const prefix = ciphertext.subarray(MAGIC.length, MAGIC.length + NONCE_PREFIX_SIZE);

  const view = new DataView(ciphertext.buffer, ciphertext.byteOffset, ciphertext.byteLength);
  const metaLen = view.getUint16(MAGIC.length + NONCE_PREFIX_SIZE, false);

  const metaEnd = HEADER_SIZE + metaLen;
  if (metaEnd > ciphertext.length) throw new DecryptError("truncated file");

  const metaJSON = await open(
    cryptoKey,
    nonce(prefix, 0, false),
    ciphertext.subarray(HEADER_SIZE, metaEnd)
  );

  let meta: E2EMetadata;
  try {
    meta = JSON.parse(new TextDecoder().decode(metaJSON));
  } catch {
    throw new DecryptError("wrong key, or the file was modified or truncated");
  }

  // Chunk layout is fully determined by the remaining length, so the final
  // chunk — and therefore the final marker — is known without a lookahead.
  const remaining = ciphertext.length - metaEnd;
  const fullChunks = Math.floor(remaining / CHUNK_CIPHER_SIZE);
  const rest = remaining % CHUNK_CIPHER_SIZE;

  if (rest !== 0 && rest < TAG_SIZE) throw new DecryptError("truncated file");

  const chunkCount = rest > 0 ? fullChunks + 1 : fullChunks;
  if (chunkCount === 0) throw new DecryptError("truncated file");

  const parts: BlobPart[] = [];
  let offset = metaEnd;

  for (let i = 0; i < chunkCount; i++) {
    const size = i === chunkCount - 1 && rest > 0 ? rest : CHUNK_CIPHER_SIZE;
    const final = i === chunkCount - 1;

    const plain = await open(
      cryptoKey,
      nonce(prefix, i + 1, final),
      ciphertext.subarray(offset, offset + size)
    );

    parts.push(plain as BlobPart);
    offset += size;
    onProgress?.(offset - metaEnd, remaining);
  }

  return { meta, blob: new Blob(parts, { type: meta.type || "application/octet-stream" }) };
}
